SI-7006 Prevent unreachable blocks in GenICode

JamesIry opened the pull request:
This commit makes GenICode prevent the generation of
most unreachable blocks. The new unreachable block prevention code can
be disabled with a compiler flag.

Because full unreachable analysis is no longer necessary for
normal code it makes the unreachable block analysis run only under
-optimise.

A test is included to make sure unreachable code doesn't cause issues
in code gen.

A concrete example will help.

  def foo(): X = {
   try
     return something()
   catch {
     case e: Throwable =>
        println(e)
        throw e
   }
   unreachableCode()
  ]

Here unreachableCode() is unreachable but GenICode would create ICode
for it and then ASM would turn it into a pile of NOPS.

A previous commit added a reachability analysis step to eliminate
that unreachable code but that added a bit of time to the
compilation process even when optimization was turned off.
This commit avoids generating most unreachable
ICode in the first place so that full reachability analysis is
only needed after doing other optimization work.

The new code works by extending a mechanism that was already in place.
When GenICode encountered a THROW or RETURN it would put the
current block into "ignore" mode so that no further instructions
would be written into the block. However, that ignore mode flag
was itself ignored when it came to figuring out if follow on blocks
should  be written. So this commit goes through places like try/catch
and if/else and uses the ignore mode of the current block to decide
whether to create follow on blocks, or if it already has, to kill by
putting them into ignore mode and closing them where they'll be
removed from the method's list of active blocks.

It's not quite as good as full reachability analysis. In particular
because a label def can be emitted before anything that jumps to it,
this simple logic is forced to leave label defs alone and that means
some of them may be unreachable without being removed. However, in
practice it gets close the the benefit of reachability analysis at
very nearly no cost.

JamesIry committed:
SI-7006 Prevent unreachable blocks in GenICode

This commit makes GenICode prevent the generation of
most unreachable blocks. The new unreachable block prevention code can
be disabled with a compiler flag.

Because full unreachable analysis is no longer necessary for
normal code it makes the unreachable block analysis run only under
-optimise.

A test is included to make sure unreachable code doesn't cause issues
in code gen.

A concrete example will help.

  def foo(): X = {
   try
     return something()
   catch {
     case e: Throwable =>
        println(e)
        throw e
   }
   unreachableCode()
  ]

Here unreachableCode() is unreachable but GenICode would create ICode
for it and then ASM would turn it into a pile of NOPS.

A previous commit added a reachability analysis step to eliminate
that unreachable code but that added a bit of time to the
compilation process even when optimization was turned off.
This commit avoids generating most unreachable
ICode in the first place so that full reachability analysis is
only needed after doing other optimization work.

The new code works by extending a mechanism that was already in place.
When GenICode encountered a THROW or RETURN it would put the
current block into "ignore" mode so that no further instructions
would be written into the block. However, that ignore mode flag
was itself ignored when it came to figuring out if follow on blocks
should  be written. So this commit goes through places like try/catch
and if/else and uses the ignore mode of the current block to decide
whether to create follow on blocks, or if it already has, to kill by
putting them into ignore mode and closing them where they'll be
removed from the method's list of active blocks.

It's not quite as good as full reachability analysis. In particular
because a label def can be emitted before anything that jumps to it,
this simple logic is forced to leave label defs alone and that means
some of them may be unreachable without being removed. However, in
practice it gets close the the benefit of reachability analysis at
very nearly no cost.

JamesIry commented:
review @gkossakowski

gkossakowski commented:
@JamesIry: shouldn't we have a test that checks whether unreachable blocks are properly eliminated?

gkossakowski commented:
Also, the commit message is not detailed enough to aid understanding of this change. @JamesIry agreed to update it.

JamesIry commented:
I'd like to avoid writing more icode tests so I'll visit ways to run a JVM test with the full reachability analysis disabled then re-visit this PR.

JamesIry commented:
Okay @gkossakowski, this should be better.

dragos commented:
Unreachability analysis would be very handy as a user-level warning, too.

JamesIry commented:
@dragos At the point this code runs it's not at all clear what was code written by the user and what was synthesized by previous stages. I'd love to have a good reachability warning for user code, but it'll take some work to prevent false warnings.

dragos commented:
@JamesIry I know, it was more of an encouragement to work on a that feature. :)

gkossakowski commented:
FYI: I'm reviewing this right now.

gkossakowski commented on src/compiler/scala/tools/nsc/backend/icode/GenICode.scala:
```diff
@@ -980,12 +990,15 @@ abstract class GenICode extends SubComponent  {
               }
 
               caseCtx = genLoad(body, tmpCtx, generatedType)
+              afterCtxReachable |= !caseCtx.bb.ignore
```
I don't know if this is nitpicking but I'm not a fan of using `|=`. Lex discussed this recently:
https://groups.google.com/d/msg/scala-language/C6ns64TphN8/6HHKor4RFrsJ

However, I won't press this too hard. Just wanted to make this observation.

JamesIry replied:
should be ||=.

gkossakowski replied:
Oh, do you say that this is side-effecting operation here?

JamesIry replied:
No, it's not. It's just that ||= is more clear than |=. The short-circuiting of the ||= over |= isn't important here because the computation on the right is both pure and cheap.

gkossakowski commented on src/compiler/scala/tools/nsc/backend/icode/GenICode.scala:
```diff
@@ -1342,9 +1355,9 @@ abstract class GenICode extends SubComponent  {
-      def genComparisonOp(l: Tree, r: Tree, code: Int) {
+                        elseCtx: Context): Boolean =
+    {      
+      def genComparisonOp(l: Tree, r: Tree, code: Int): Boolean = {
```
We would benefit from a comment that explains what return value means here.

JamesIry replied:
Sure

gkossakowski commented on src/compiler/scala/tools/nsc/backend/icode/GenICode.scala:
```diff
@@ -1436,7 +1456,7 @@ abstract class GenICode extends SubComponent  {
      * @param elseCtx target context if the comparison yields false
      */
-    def genEqEqPrimitive(l: Tree, r: Tree, ctx: Context)(thenCtx: Context, elseCtx: Context): Unit = {
+    def genEqEqPrimitive(l: Tree, r: Tree, ctx: Context)(thenCtx: Context, elseCtx: Context): Boolean = {
```
Document return value, please.

gkossakowski commented on test/files/jvm/unreachable/Test.scala:
```diff
@@ -0,0 +1,23 @@
+  def show: Unit = {
+    val classNode = loadClassNode("Foo_1")
+    // Foo_1 is full of unreachable code which if not elimintated
+    // will result in NOPs as can be confirmed by adding -Ydisable-unreachable-prevention
```
I'm wondering if more direct way by looking for magic constants in unreachable code would be better. If we really optimize here for lack of NOPs then this test case is fine.

JamesIry replied:
Not really, when ASM finds unreachable code while building the stack map it turns it into NOPs. That's because unreachable code needs stack map entries but unreachable ordinary instructions can't be typed since the incoming type state is undetermined. So ASM NOPs them which is easy to type.

gkossakowski commented:
@JamesIry: definitively an improvement of previous iteration. The core of your patch looks good to me already.

I'd like to hear from you an opinion on points I raised.

JamesIry committed:
SI-7006 Cleanup from code review

Minor cleanup from review of https://github.com/scala/scala/pull/2185

* Changed several instances of |= to ||= for better clarity (and bonus
savings!)
* Documented the return of two methods that compute the reachability
of follow-on blocks.

gkossakowski commented:
LGTM.
